### ISSUE TITLE
THPVariable_subclass_dealloc (i.e. tp_dealloc for torch.Tensor) should handle Python error state gracefully

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#65083]
## 🐛 Bug

Identified by @albanD, from offline discussion with @ezyang: "You cannot assume that there isn't an error already set when THPVariable_dealloc is called.
If you need to call some functions that might clobber the global error state in there, we should PyErr_Fetch at the beginning of the function and PyErr_Restore at the end."

## To Reproduce

See https://github.com/pytorch/pytorch/pull/65068

## Expected behavior

We should call PyErr_Fetch at the beginning of THPVariable_subclass_dealloc and PyErr_Restore at the end. However (and this is the reason why I haven't submitted a patch for this yet) we have to be very, very careful about what happens if a Python error (call this error B) occurs in between PyErr_Fetch and PyErr_Restore. I guess we have a few options here:
- completely ignore error B
- don't PyErr_Restore if error B occured
- PyErr_Restore only if an error was previously set before entering tp_dealloc
Does python have a better solution for this?

Here's some examples of this out in the wild:
- https://github.com/posborne/dbus-python/blob/fef4bccfc535c6c2819e3f15384600d7bc198bc5/_dbus_bindings/conn.c#L387
- https://bugs.python.org/issue45210?%40ok_message=msg%20401854%20created%0Aissue%2045210%20created&%40template=item

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @bhosmer @smessmer @ljk53

**Comment by zou3519** [comment]
[This thread](https://stackoverflow.com/questions/52468416/who-should-call-pyerr-fetch) suggests that PyErr_WriteUnraisable should be used if an exception happens in between Fetch and Restore

**Comment by zou3519** [comment]
I'm not sure this needs to be high-pri (I'm not sure how to write python code to trigger the bug; no users have come to us about this yet) but the fix is not too difficult.

**Comment by albanD** [comment]
I think that printing the ignored exception and then restore the previous exception state is the simple way to go here.

Also I am not so sure this cannot be triggered in some weird edge case.
In any case, it leads to very hard-to-debug issue when modifying Tensor methods so I think we should fix it.

**Comment by zou3519** [comment]
emperically not hi-pri if open for years

**Comment by albanD** [comment]
Agreed, also actionable as the next step seemed obvious even though I don't recall all the subtlety here.
